File: 2685/solution.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        paths = defaultdict(list)
        for u,v in edges:
            paths[u].append(v)
            paths[v].append(u)
        vis = set()
        def dfs(node):
            vis.add(node)
            nodeCnt = 1
            edgeCnt = len(paths[node])
            for nei in paths[node]:
                if nei not in vis:
                    nd, ed = dfs(nei)
                    nodeCnt+=nd
                    edgeCnt+=ed 
            return nodeCnt, edgeCnt
        res = 0
        for i in range(n):
            if i in vis:
                continue
            m, e = dfs(i)
            e//=2 
            if e==m*(m-1)//2:
                res+=1
        return res

File: 2685/solution.java
class Solution {

    Map<Integer, List<Integer>> paths = new HashMap<>();
    Set<Integer> vis = new HashSet<>();

    public int[] dfs(int node){
        vis.add(node);
        int nodes = 1;
        int edges = paths.getOrDefault(node, new ArrayList<>()).size();
        for(int nei: paths.getOrDefault(node, new ArrayList<>())){
            if(!vis.contains(nei)){
                int[] nodesAndEdges = dfs(nei);
                nodes+=nodesAndEdges[0];
                edges+=nodesAndEdges[1]; 
            }
        }
        return new int[] {nodes, edges};
    }

    public int countCompleteComponents(int n, int[][] edges) {
        int res = 0;
        for(int[] edge: edges){
            int src = edge[0];
            int dest = edge[1];
            paths.computeIfAbsent(src, k -> new ArrayList<>()).add(dest);
            paths.computeIfAbsent(dest, k -> new ArrayList<>()).add(src);
        }

        for(int i=0;i<n;i++){
            if(vis.contains(i))
                continue;
            int[] ne = dfs(i);
            ne[1]/=2;
            if(ne[1]==ne[0]*(ne[0]-1)/2)
                res++;
        }
        return res;
    }
}
